sklearn/pipeline.py

```diff
@@ -831,3 +841,114 @@ def make_union(*transformers, **kwargs):
+
+    Parameters
+    ----------
+    transformers : list of tuples
+        List of (name, transformer, column) tuples specifying the transformer
+        objects to be applied to subsets of the data. The columns can be
+        specified as a scalar or list (for multiple columns) of integer or
+        string values. Integers are interpreted as the positional columns,
+        strings as the keys of `X`. In case of positional integers, also
+        a slice object is accepted.
+
+    n_jobs : int, optional
+        Number of jobs to run in parallel (default 1).
+
+    transformer_weights : dict, optional
+        Multiplicative weights for features per transformer.
+        Keys are transformer names, values the weights.
+
+    Examples
+    --------
+    >>> from sklearn.pipeline import ColumnTransformer
+    >>> from sklearn.preprocessing import Normalizer
+    >>> union = ColumnTransformer(
+    ...     [("norm1", Normalizer(norm='l1'), 'subset1'),
+    ...      ("norm2", Normalizer(norm='l1'), 'subset2')])
+    >>> X = {'subset1': [[0., 1.], [2., 2.]], 'subset2': [[1., 1.], [0., 1.]]}
+    >>> union.fit_transform(X)    # doctest: +NORMALIZE_WHITESPACE
+    array([[ 0. ,  1. ,  0.5,  0.5],
+           [ 0.5,  0.5,  0. ,  1. ]])
+
+    """
+    def _iter(self, X=None, skip_none=True):
+        """Generate (name, trans, column, weight) tuples
+        """
+        get_weight = (self.transformer_weights or {}).get
+        return ((name, trans, _getitem(X, column), get_weight(name))
+                for name, trans, column in self.transformer_list
+                if not skip_none or trans is not None)
+
+    def _update_transformer_list(self, transformers):
+        transformers = iter(transformers)
+        self.transformer_list[:] = [
+            (name, None if old is None else next(transformers), column)
+            for name, old, column in self.transformer_list
+        ]
+
+
+def _getitem(X, column):
+    """
+    Get feature column from input data (array, dataframe, dict)
+
+    """
+    if X is None:
+        return X
+
+    # check whether we have string column names or integers
+    if (isinstance(column, int) or isinstance(column, slice)
+            or (isinstance(column, list)
+                and all(isinstance(col, int) for col in column))):
+        column_names = False
+    elif (isinstance(column, six.string_types)
+          or (isinstance(column, list)
+              and all(isinstance(col, six.string_types) for col in column))):
+        column_names = True
+    else:
+        raise ValueError("no valid 'column' type")
+
+    if column_names:
+        if hasattr(X, 'loc'):
+            # pandas dataframes
+            if not isinstance(column, list):
+                column = [column]
+            return X.loc[:, column]
+        else:
+            # dicts or numpy recarrays
+            if not isinstance(column, list):
+                return _ensure_2d(X[column])
```

Comment:
Yes (this was one of the todo's). See the comments in the main thread  (https://github.com/scikit-learn/scikit-learn/pull/9012#issuecomment-306825163 and below). What you suggest is basically the option 4 I think.

Apart from the shape (1D vs 2D) problem, there is also the problem of the conversion to an array in case of text data (eg when you have a dict -> list of text data). Just converting to an array is not what we want, because this will convert it to some numpy string dtype. But to do this well, you need to start inferring the type of the data in the list, leading to a whole new can of worms .. (a problem that is also encountered in https://github.com/scikit-learn/scikit-learn/pull/8793 to support strings in the OneHotEncoder)

If we don't reshape X to a 2D array in case of a scalar column selection, we could in principle leave such lists intact, avoiding this problem. 
But you would still have this conversion problem for text columns when selecting multiple columns.